lib/Runtime/Types/RecyclableObject.h

```diff
@@ -482,36 +496,41 @@ namespace Js {
         bool isRecyclableObject = (((uintptr_t)aValue) & AtomTag) == AtomTag_Object;
 #endif
 
-        return isRecyclableObject && VarIs<T>(reinterpret_cast<RecyclableObject*>(aValue));
+        return isRecyclableObject && VarIsImpl<T>(reinterpret_cast<RecyclableObject*>(aValue));
     }
 
-    template <> inline bool VarIs<RecyclableObject>(RecyclableObject* obj) { return true; }
+    // Validate that the object is actually the type that the type system thinks it is.
+    // This should only be used for extremely defensive assertions; if you find code
+    // relying on this behavior for correctness, then it's cause for concern.
+    template <typename T> bool VarIsCorrectType(T* obj)
```

Comment:
This feels like there's a few situations in which a smart compiler would be able to make it always return true - e.g. with a static build and ltcg, I think it may be legal to optimize the "GetTypeId() == mytypeid" ::Is() checks for various object types in our type system to always return true when working on the same type as that passed in - since there's no avenue other than UB to change the typeid of valid objects of that pointer type to something else. Since we're already doing effectively that, it's not a regression with this change, but in a future change we'll most likely want to change this to being closer to VarIs in terms of signature, with the exception of making the last line `AssertOrFailFast(VarIsImpl<t>(obj))`.